File: src/llm_usage/identity.py
from __future__ import annotations

import hashlib
from typing import Optional


def hash_user(username: str, salt: str) -> str:
    payload = f"{username}|{salt}".encode("utf-8")
    return hashlib.sha256(payload).hexdigest()


def hash_source_host(username: str, source_label: str, salt: str) -> str:
    payload = f"{username}|{source_label}|{salt}".encode("utf-8")
    return hashlib.sha256(payload).hexdigest()


def build_row_key(
    user_hash: str,
    source_host_hash: str,
    date_local: str,
    tool: str,
    model: str,
    session_fingerprint: Optional[str] = None,
) -> str:
    identity = session_fingerprint.strip() if session_fingerprint and session_fingerprint.strip() else f"model:{model}"
    payload = f"{user_hash}|{source_host_hash}|{date_local}|{tool}|{identity}".encode("utf-8")
    return hashlib.sha256(payload).hexdigest()

File: node/src/core/identity.js
import { sha256Hex } from "./hash.js";

export function hashUser(username, salt) {
  return sha256Hex(`${username}|${salt}`);
}

export function hashSourceHost(username, sourceLabel, salt) {
  return sha256Hex(`${username}|${sourceLabel}|${salt}`);
}

export function resolveIdentity(model, sessionFingerprint) {
  if (typeof sessionFingerprint === "string" && sessionFingerprint.trim()) {
    return sessionFingerprint.trim();
  }
  return `model:${model}`;
}

export function buildRowKey({
  userHash,
  sourceHostHash,
  dateLocal,
  tool,
  model,
  sessionFingerprint = null,
}) {
  const identity = resolveIdentity(model, sessionFingerprint);
  return sha256Hex(`${userHash}|${sourceHostHash}|${dateLocal}|${tool}|${identity}`);
}
